omni-relayer/src/startup/solana.rs

```diff
@@ -0,0 +1,343 @@
+    };
+
+    let Ok((mut log_stream, _)) = ws_client.logs_subscribe(filter, config).await else {
+        anyhow::bail!("Failed to subscribe to Solana logs");
+    };
+
+    info!("Subscribed to live Solana logs");
+
+    while let Some(log) = log_stream.next().await {
+        let Ok(signature) = Signature::from_str(&log.value.signature) else {
+            warn!("Failed to parse signature: {:?}", log.value.signature);
+            continue;
+        };
+
+        info!("Processing signature: {:?}", signature);
+
+        // TODO: We need to wait for the transaction to be confirmed, but this must be replaced
+        // with a worker
+        tokio::time::sleep(tokio::time::Duration::from_secs(15)).await;
+
+        match http_client
+            .get_transaction(&signature, UiTransactionEncoding::Json)
+            .await
+        {
+            Ok(tx) => {
+                let transaction = tx.transaction;
+
+                if let solana_transaction_status::EncodedTransaction::Json(ref tx) =
+                    transaction.transaction
+                {
+                    if let UiMessage::Raw(ref raw) = tx.message {
+                        process_message(
+                            &mut redis_connection,
+                            &solana,
+                            &transaction,
+                            raw,
+                            signature,
+                        )
+                        .await
+                    }
+                }
+
+                utils::redis::update_last_processed_block(
+                    &mut redis_connection,
+                    &utils::redis::get_last_processed_block_key(ChainKind::Sol).await,
+                    tx.slot,
+                )
+                .await;
+            }
+            Err(e) => {
+                warn!("Failed to fetch transaction: {}", e);
+            }
+        };
+    }
+
+    Ok(())
+}
+
+async fn process_recent_logs(
+    redis_connection: &mut redis::aio::MultiplexedConnection,
+    solana: &config::Solana,
+    http_client: &RpcClient,
+    program_id: &Pubkey,
+) -> Result<()> {
+    let Some(start_block_height) = utils::redis::get_last_processed_block(
+        redis_connection,
+        &utils::redis::get_last_processed_block_key(ChainKind::Sol).await,
+    )
+    .await
+    else {
+        return Ok(());
+    };
+
+    // TODO: Replace all this simply by saving last signature instead of slot
+    let mut last_signature = http_client
+        .get_signatures_for_address_with_config(
+            program_id,
+            GetConfirmedSignaturesForAddress2Config {
+                limit: Some(1000),
+                before: None,
+                until: None,
+                commitment: Some(CommitmentConfig::confirmed()),
+            },
+        )
+        .await?
+        .into_iter()
+        .find(|sig| sig.slot == start_block_height)
+        .and_then(|sig| Signature::from_str(&sig.signature).ok());
+
+    if last_signature.is_none() {
```

Comment:
What if the first 1000 aren’t enough and the required transaction is further back?

Reply:
https://github.com/Near-One/omni-bridge/pull/153#discussion_r1892662768

Reply:
Fixed in https://github.com/Near-One/omni-bridge/pull/153/commits/76f9138a06ed9846d94dc29fa9ac18e1ac57a2ec